[GR-42601] Model class constants in debug info

adinn opened the pull request:
This PR exposes the location of java.lang.Class (aka DynamicHub) instances located in the image heap to debug info generators, allowing them to be encoded into debug info using the standard class name as an identifier for the corresponding object.

It includes a modification to the Linux debug info generator that encodes details of the class constant object in the info section as a DWARF constant DIE. The implementation enables reference to the class constants using symbolic names like

```
  hello.Hello.class
  java.util.List.class
  java.long.String[].class
  long.class
  byte[].class
```

The type of all the above constants is `class java.lang.Class` (i.e. they identify the underlying object layout rather than an oop).

These symbols are not currently entered into the symbol table as local symbols. That might be desirable for a follow up PR.

adinn commented:
This is being proposed as a solution for issue #5251.

fniephaus commented:
Thanks @adinn! This LGTM but I'll let the others comment on the implementation.

BTW: Running `mx checkstyle --primary` in `substratevm/` before pushing has saved me a lot of time fighting with the CI 😅.

zakkak reviewed (approved):
LGTM, thanks Andrew!

zakkak commented:
> BTW: Running `mx checkstyle --primary` in `substratevm/` before pushing has saved me a lot of time fighting with the CI sweat_smile.

Sometimes even that is not enough and I need to run the following as well:

```
mx --primary-suite-path substratevm --J @-Xmx2g eclipseformat -e ~/.mx/eclipse/eclipse --primary
```

adinn commented:
@fniephaus @zakkak I ran both checkstyle and eclipseformat but the local runs don't always show all the errors that turn up in the gate checks.

I have pushed further style fixes. Let's hope this one catches everything.

fniephaus commented:
> I have pushed further style fixes. Let's hope this one catches everything.

Thanks! Looks like this has worked :)

olpaw commented on docs/reference-manual/native-image/DebugInfo.md:
````diff
@@ -394,6 +395,46 @@ end
+classes and arrays, including primitive arrays:
+
+```
+(gdb)  print 'java.util.List.class'.name
````
@adinn, an alternative approach would be to add an artificial static `class` field to each type so that the hub of a given type can be accessed via
```
p *'java.util.List'::class
```
wouldn't that fit better to the existing code C++ based code model that we have?

peter-hofer replied:
I believe the field would better be called `.class` so that it's not a valid Java bytecode field name, if DWARF permits that and it doesn't make problems with GDB, that is:
`'java.util.List'::.class`

adinn replied:
> @adinn, an alternative approach would be to add an artificial static class-field to each type so that the hub of a given type can be accessed via
> ```
> p *'java.util.List'::class
> ```
> wouldn't that fit better to the existing C++ based code model that we have?

@olpaw That is indeed a nicer fit. The only problem with it is that it doesn't work ;-).

The only real option we have to make this data value visible via debug info is to encode it as a variable/constant, whether at top level or as a (static) member of some class. Any corresponding info definition must identify where the variable is located in memory. So, a static field info definition for a pseudo-field as described above would need specify as location the address of a memory slot whose content is an `oop`, a reference to the `Class` instance. We can make that reference a direct or indirect( (encoded) address by using a suitable location expression. However, we cannot get rid of the need for there to be an indirection.

Unfortunately, the only thing we have available to encode as the location is the address of the `Class` instance. *That* is the reason why I have had to declare the type of constant `java.util.List.class` as `java.lang.Class` not `java.lang.Class *`. It's also why when using isolates the type remains `java.lang.Class` not `_z_.java.lang.Class` or `_z_.java.lang.Class *`

That said, it is still possible to embed this constant as a field of the class. The required change is simply to generate the declaration and definition with the scope of the class and adjust the name and linkage name. The inescapable constraint we are under is that the field has to be an object value not an object reference.

So, we can support either:
```
(gdb) p 'java.util.List'::class
```
or
```
(gdb) p 'java.util.List'::.class
```

n.b. the reason I chose to make the constant a top level value rather than a value embedded in the class was:

1. the syntax and use of name `java.util.List.class` is very much like that of its Java original i.e. it functions as a global name for the class (modulo the associated network of class loader spaces within which it is visible)
2. in Java the class object is not actually a field of the class so presenting it as such appeared to me to risk engendering unnecessary confusion

I'm very happy to change this if you still want me to after taking the above explanation into account.

@peter-hofer yYour worry about `Foo::class` clashing with a legitimate field name is unnecessary. `class` is a reserved keyword so it cannot be used as a field name.

peter-hofer replied:
@adinn `class` is a keyword in the Java _language_, but I would be surprised if it's explicitly forbidden in Java bytecode and can therefore not be used by other languages and especially bytecode generating tools.

adinn replied:
@peter-hofer Oh yes, of course `class` could well be inserted into synthetic or transformed bytecode as a class pool constant. I'd discounted that as something to worry about but now that you mention it I suppose there is quite probably someone somewhere crazy enough to have decided to use it as as field name.

Mind you, the same concern would apply for `.class`, perhaps yet more so in that it is not a keyword and therefore more attractive to less crazy people ;-).

I actually think this is all the more reason to keep encoding the class constants as a top level name. All other Java names are going to exist within the scope of a class (which effectively means a namespace). I think it is relatively unlikely that something like `org.my.Foo.class` is going to turn up as a name for a variable in non-Java code. Of course, that is asking to be proven wrong ...

peter-hofer replied:
@adinn `.class` is not a valid identifier in bytecode because `.` is not allowed in identifiers (JVMS 4.2.2), hopefully that rule is enforced too. `org.my.Foo.class` could again refer to a qualified class or interface named `class` in package `org.my.Foo` -- hopefully it never does, but it's not prohibited.

olpaw replied:
@peter-hofer is right, it's trivial to end up with names that are reserved keywords in Java with bytecode producers other than `javac`. E.g.
```scala
object Hello {
    val `class`: String = "WHY NOT?"
    def main(args: Array[String]) = {
        println(`class`)
    }
}
```
Unfortunately I'm not sure about `.class` either, because the following also works just fine:
```scala
object Hello {
    val `.class`: String = "WHY NOT?"
    def main(args: Array[String]) = {
        println(`.class`)
    }
}
```
But it could be that here Scala applies some magic to make this possible.

peter-hofer replied:
@olpaw, Scala seems to do just that because the field is actually named `$u002Eclass` when using reflection.

peter-hofer replied:
```
object Hello {
  val `class`: String = "OH NO";
  val `.class`: String = "WHY NOT?";

  def main(args: Array[String]) = {
    for (f <- getClass.getDeclaredFields) {
      println(f.getName())
    }
  }
}
```
prints
```
class
$u002Eclass
MODULE$
```

adinn commented:
@olpaw I have fixed up the commit messages and squashed the style fixes.
@peter-hofer  @vvlevchenko Just waiting for your decision on whether to continue with the current use of a top level var `org.my.Foo.class` var or switch to a static field `org.my.Foo::class`?

peter-hofer commented:
@adinn, could the top-level variable alias with anything else (a type)? If not, the current solution seems fine to me.

adinn commented:
> @adinn, could the top-level variable alias with anything else (a type)? If not, the current solution seems fine to me.

@peter-hofer Yes, there is potential for such a name clash in the scenario you described i.e. we have a Java class `org.my.Foo` and also encounter a bytecode file crafted to define a class named `class` in package `org.my.Foo`. Then with my proposal `org.my.Foo.class` would be the name for both the class constant and the hand-crafted object type. If we adopt your suggestion then there is no possibility of a clash.

adinn commented:
@peter-hofer Hmm, I was looking into what code changes would be needed to switch to implementing class constants using static field syntax (e.g. `'org.my.Foo':.class`) and ran into a problem. There is no direct equivalent for primitive type class constants because we don't (can't) generate a corresponding info class DIE. The names for the Java primitive types must used to declare primitive types so that they can then be referenced from class definitions. So, we cannot also have a class with that same name.

We might be able to get round this by embedding the class constant for primitive type in a namespace e.g. for `int.class` we create a namespace called `int` and declare a variable called `.class` in that namespace. I'm not sure whether that is going to create a problem with `int` doing double duty as the identifier for a primitive type and a namespace.

adinn commented:
@olpaw ^^^ see above comment

olpaw commented:
Approved. Sorry for the late response!

peter-hofer commented:
Perhaps there is GDB syntax to distinguish a global variable `org.my.Foo.class` from a type of the same name, too. I do hope that this never occurs in practice anyway though.

olpaw commented:
Running in our CI now

olpaw commented:
Merged as https://github.com/oracle/graal/pull/5478
